packages/core/src/compiler/styleFunctionCaller/StyleFunctionCaller.ts

```diff
@@ -0,0 +1,101 @@
+import * as ad from "../../types/ad";
+import { SourceRange } from "../../types/ast";
+import {
+  CompFunc,
+  ConstrFunc,
+  FuncParam,
+  ObjFunc,
+} from "../../types/functions";
+
+import { Result } from "true-myth";
+import { Context } from "../../shapes/Samplers";
+import { Shape } from "../../shapes/Shapes";
+import { StyleError } from "../../types/errors";
+import { ArgValWithSourceLoc, Value } from "../../types/value";
+import {
+  badArgumentTypeError,
+  functionInternalError,
+  missingArgumentError,
+  tooManyArgumentsError,
+} from "../../utils/Error";
+import { checkType } from "../styleTypeChecker/styleTypeChecker";
+
+const { ok, err } = Result;
+export const callCompFunc = (
+  func: CompFunc,
+  range: SourceRange,
+  context: Context,
+  args: ArgValWithSourceLoc<ad.Num>[]
+): Result<Value<ad.Num>, StyleError> => {
+  if (args.length > func.params.length) {
+    return err(tooManyArgumentsError(func, range, args.length));
+  }
+  const vals: (Shape<ad.Num> | Value<ad.Num>["contents"])[] = [];
+  for (let i = 0; i < func.params.length; i++) {
+    const funcArg = func.params[i];
+    const arg: ArgValWithSourceLoc<ad.Num> | undefined = args[i];
+    const v = checkArg(func.name, range, funcArg, arg);
+    if (v.isErr()) return err(v.error);
+    vals.push(v.value);
+  }
+  try {
+    return ok(func.body(context, ...vals));
+  } catch (e) {
+    if (e instanceof Error) {
+      return err(functionInternalError(func, range, e.message));
+    } else {
+      throw new Error("Function call resulted in exception not of Error type");
+    }
+  }
+};
+
+export const callObjConstrFunc = (
+  func: ObjFunc | ConstrFunc,
+  range: SourceRange,
+  args: ArgValWithSourceLoc<ad.Num>[]
+): Result<ad.Num, StyleError> => {
+  if (args.length > func.params.length) {
+    return err(tooManyArgumentsError(func, range, args.length));
+  }
+  const vals: (Shape<ad.Num> | Value<ad.Num>["contents"])[] = [];
+  for (let i = 0; i < func.params.length; i++) {
+    const funcArg = func.params[i];
+    const arg: ArgValWithSourceLoc<ad.Num> | undefined = args[i];
+    const v = checkArg(func.name, range, funcArg, arg);
+    if (v.isErr()) return err(v.error);
+    vals.push(v.value);
+  }
+  try {
+    return ok(func.body(...vals));
+  } catch (e) {
+    if (e instanceof Error) {
+      return err(functionInternalError(func, range, e.message));
+    } else {
+      throw new Error("Function call resulted in exception not of Error type");
+    }
+  }
+};
```

Comment:
I'd suggest de-duplicating this code a bit. Seems like only the context-passing bit is different. You may extract the rest of the function into `checkArgs` and make this much shorter.

Reply:
Sounds good, I'll see how I can refactor this.

Reply:
Done